python_modules/dagster/dagster/_core/definitions/metadata/source_code.py

```diff
@@ -0,0 +1,152 @@
+import inspect
+import os
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Mapping,
+    Optional,
+    Sequence,
+    Union,
+)
+
+import pydantic
+
+import dagster._check as check
+from dagster._annotations import experimental
+from dagster._model import DagsterModel
+from dagster._serdes import whitelist_for_serdes
+
+from .metadata_set import (
+    NamespacedMetadataSet as NamespacedMetadataSet,
+    TableMetadataSet as TableMetadataSet,
+)
+from .metadata_value import MetadataValue
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.assets import AssetsDefinition, SourceAsset
+    from dagster._core.definitions.cacheable_assets import CacheableAssetsDefinition
+
+DEFAULT_SOURCE_FILE_KEY = "asset_definition"
+
+
+@experimental
+@whitelist_for_serdes
+class LocalFileCodeReference(DagsterModel):
+    """Represents a local file source location."""
+
+    file_path: str
+    line_number: int
+
+
+@experimental
+@whitelist_for_serdes
+class CodeReferencesMetadataValue(DagsterModel, MetadataValue["CodeReferencesMetadataValue"]):
+    """Metadata value type which represents source locations (locally or otherwise)
+    of the asset in question. For example, the file path and line number where the
+    asset is defined.
+
+    Attributes:
+        sources (Dict[str, LocalFileCodeReference]):
+            A labeled dictionary of sources. The main source file should be keyed with
+            the `DEFAULT_SOURCE_FILE_KEY` constant.
+    """
+
+    sources: Mapping[str, LocalFileCodeReference] = {}
```

Comment:
Questions:

1) What is the key, exactly?
2) Shouldn't this be `code_references` now?

Reply:
Updated the PR to drop the JS changes for review-ability - they're needed to get tests passing, but are all generated and can just re-generated before the merge.

1. Here, the key is a user-specified label for the code reference in the case that there is more than one for a particular asset. For most assets, there is only a single code reference: the source file where the `@asset` annotated function lives, and it is keyed with the `DEFAULT_SOURCE_FILE_KEY` constant.

  There are some cases where users may want to point to multiple files. From a DBT model, our integration may want to link to the Dagster `@dbt_assets` definition as well as the `.sql` model file. For a YAML-defined asset using some sort of user-defined YAML layer, they may want to link to both the Python code defining the asset and the YAML specification.

  See #21468 for how this looks in the UI, though we are going to do another design pass.

2. Yes, updated.